### DISCUSSION TITLE
feat: add destroyed town-hall cost, require materials and stop city growth (until rebuilt)

### DISCUSSION THREAD

**Discussion by winex** [bahrmichael/factorio-tycoon#287] [Ideas]
https://github.com/bahrmichael/factorio-tycoon/blob/fbbe87d273a89e188cd8255b870af680d61f168e/construction-event-handler.lua#L82-L87

just got an idea about this - i propose we:
- make `town-hall` "constructable" (not players) in the first place, so it needs to be finished in `completeConstruction()`
- stop any growing (or all!) activity for this city if `town-hall` was destroyed
- require some materials in HW stores to "reconstruct" `town-hall`
_stores might be destroyed by biters too, but they're already buildable / player-craftable_

plus related:
- ? change roads to: dirt roads(???) -> stone bricks -> concrete (looks weird if no research and w/o such materials)
- increase initial city grid to *2x2* empty "cells", so we have a real-looking center, with a crossroad at least :)
- randomly place `town-hall` (and treasury probably) in one of that 4 cells

had some thoughts about city center vs `town-hall` as city center... city centers in real world are mostly thought as road Squares, not some building - so the tags are now be placed correctly, not at `town-hall` position.
proposed changes above will emphasize this...

**Comment by bahrmichael** [comment]
> make town-hall "constructable" (not players) in the first place, so it needs to be finished in completeConstruction()

That's on my immediate to-do list :) There will be a recipe that we can use, and players can unlock the recipe when they research Multiple Cities. I plan on removing the urban planning center and its mechanics.

> stop any growing (or all!) activity for this city if town-hall was destroyed

We can do that, if we allow it to be rebuilt early in the game.

> require some materials in HW stores to "reconstruct" town-hall

I'm not sure about that yet. I plan on making the recipe relatively expensive. That means it would need a lot of space in hardware stores, and also how do we communicate to players what they need to rebuild it? This tbh feels more like a chore than nice gameplay.

**Reply by winex** [reply]
the point i was talking about - the `destroyed-by-biters` reason in the above `on_removed()` code, this could be independent of `crafting-by-player` and totally apart from `Multiple Cities` tech.
- `city-constructable` meant to be processed in `City::` class, **not by players**
- buildable by players is being `craftable` or `placeable`

> I'm not sure about that yet. I plan on making the recipe relatively expensive.

i mean `recipe != constructionCost`, the latter specified in `Constants.CONSTRUCTION_MATERIALS` and taken from HW stores to complete, while expensive recipe ingredients must be put in player or machine's inventories instead:
- HW stores <-- `CONSTRUCTION_MATERIALS[name]`
- inventory <-- `recipe[name].ingredients`

we're talking about 2 different things:
1. city with town-hall **being invalid** (destroyed, ghost, w/e), processing-wise from code perspective
2. craftable/replacable `town-hall` and it's `entity-ghost` if destroyed

**Comment by bahrmichael** [comment]
> ? change roads to: dirt roads(???) -> stone bricks -> concrete (looks weird if no research and w/o such materials)

That sounds like a nice idea! The city could then upgrade roads as it upgrades housing or as the city grows in size. Do you know of any tiles we can use as dirt roads?

**Reply by winex** [reply]
currently i know only `dirt-{1..7}` and `landfill` tiles - we can try these ones if that looks any better.
**edit:** found `dry-dirt`, looks great for roads. together with `landfill` under buildings should make map look better ;)

some mods adds more road types like `asphalt` from `pyanodon`, but i think that overall cross-mod interoperability should be avoided as much as possible, especially for `entities`, while `tiles` is the least problematic: i have some experience of **installing and removing some mods** in a single mp save -- **it doesn't go very well** and sometimes leads to mod-crashes

**Comment by bahrmichael** [comment]
> increase initial city grid to 2x2 empty "cells", so we have a real-looking center, with a crossroad at least :)
> randomly place town-hall (and treasury probably) in one of that 4 cells

I think I like the current design. Is there any reason to change this apart from asthetics?

**Reply by winex** [reply]
only 1 minor one, which cause i still couldn't get: `city.roadEnds = { first = 6, last = 5 }` https://github.com/bahrmichael/factorio-tycoon/issues/186#issuecomment-1986058243 and we should fix that first, ofc :)
